host/fde_key_manager/fde_key_manager.c

```diff
@@ -0,0 +1,692 @@
+    return ret;
+}
+
+/**
+ * There are two fde hooks, 'fde-reveal-key' and 'fde-setup'
+ * 'fde-setup' hook can be called in 2 scenarios, as true snap hook
+ * and as hook within initranfs. Depending on the runtime environment, hook
+ * uses stdin/stdout or `snapctl` as way to retrieve input and pass the result.
+ * `snapctl fde-setup-request` and `snapctl fde-setup-result` are used in snap hook
+ * runtime environment.
+ * Input and output are always passed as formated json.
+ *
+ * - 'fde-reveal-key':
+ *   - supported operations:
+ *     - reveal:
+ *         - request: { "op": "reveal", "sealed-key": "base64-encoded-bytes",
+ *                 "handle": "base64-encoded-bytes"}
+ *         - result: {"key": "base64-encoded-bytes"}
+ *     - lock:
+ *         - request: { "op": "lock" }
+ *         - result:
+ *
+ * - 'fde-setup':
+ *   - supported operations:
+ *     - fde-setup:
+ *       - request: {"op": "initial-setup","key": "base64-encoded-bytes",
+ *              "key-name" : "string"}
+ *       - result: {"sealed-key": "base64-encoded-bytes",
+ *               "handle": "base64-encoded-bytes"}
+ *       - request: {"op":"features"}
+ *       - result: {"features": []}
+ */
+int main(int argc, char *argv[]) {
+    int ret = EXIT_SUCCESS;
+    uint32_t lock;
+    unsigned char *buf = NULL;
+    char *base64_buf = NULL;
+    char *baseName = NULL;
+    char *request_str = NULL;
+    size_t buf_len = 0;
+    baseName = basename(argv[0]);
+    ree_log(REE_INFO, "main: %s entry point", baseName);
+    if (argc == 1) {
+        // handle hook run scenarios
+        // depending on hook type, we pass result on stdout or to snapctl
+        if (!strncmp(baseName,
+                              FDE_SETUP,
+                              strlen(FDE_SETUP))) {
+            // check if hook is running within initramfs or as a snap hook
+            // there is not a perfect way to do this, but certain files should
+            // only exist within initrd context
+            if (access("/etc/initrd-release", F_OK) == 0) {
+                request_str = get_initrd_fde_request();
+                snapctl_output = 0;
+            } else {
+                request_str = get_snap_hook_fde_setup_request();
+                snapctl_output = 1;
+            }
+        } else if (!strncmp(baseName,
+                            FDE_REVEAL_KEY,
+                            strlen(FDE_REVEAL_KEY))) {
+            request_str = get_initrd_fde_request();
+            snapctl_output = 0;
+        } else {
+            ree_log(REE_ERROR, "Unknown invocation" );
+            print_help();
+            exit(0);
+        }
+        if (request_str) {
+            ret = handle_fde_operation(request_str);
+            free(request_str);
+        } else {
+            ree_log(REE_ERROR, "main: empty request string");
+            ret = EXIT_FAILURE;
+        }
+        // handle systemd bug when returning too fast can take down systempd, ups
+        usleep(500000);
```

Comment:
Is there any systemd bug for this, or a link somewhere that explains the issue with a bit more of detail?

Reply:
https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1921145
but this seems to be fixed upstream so we should remove this, it will save us 0.5 seconds :)